File: java/svm_classify.java
import java.io.*;
import java.util.*;

class svm_classify {
	private static double atof(String s)
	{
		return Double.valueOf(s).doubleValue();
	}

	private static int atoi(String s)
	{
		return Integer.parseInt(s);
	}

	private static void classify(BufferedReader input, DataOutputStream output, svm_model model) throws IOException
	{
		int correct = 0;
		int total = 0;

		while(true)
		{
			String line = input.readLine();
			if(line == null) break;

			StringTokenizer st = new StringTokenizer(line," \t\n\r\f:");

			double label = atof(st.nextToken());
			int m = st.countTokens()/2;
			svm_node[] x = new svm_node[m];
			for(int j=0;j<m;j++)
			{
				x[j] = new svm_node();
				x[j].index = atoi(st.nextToken());
				x[j].value = atof(st.nextToken());
			}
			double decision_value = svm.svm_classify(model,x);
			output.writeBytes(decision_value+"\n");
			if(decision_value*label > 0)
				++correct;
			++total;
		}
		System.out.print("correct/total = "+correct+"/"+total
				+" ("+(double)correct/total*100+"%)\n");
	}

	public static void main(String argv[]) throws IOException
	{
		if(argv.length != 3)
		{
			System.err.print("usage: svm-classify test_file model_file output_file\n");
			System.exit(1);
		}

		BufferedReader input = new BufferedReader(new FileReader(argv[0]));
		DataOutputStream output = new DataOutputStream(new FileOutputStream(argv[2]));
		svm_model model = svm.svm_load_model(argv[1]);
		classify(input,output,model);
	}
}

File: svm-classify.c
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include "svm.h"
#define MAX_LINE_LEN 100000
#define MAX_NR_ATTR 1000

char line[MAX_LINE_LEN];
struct svm_node x[MAX_NR_ATTR];

struct svm_model* model;

void classify(FILE *input, FILE *output)
{
	int correct = 0;
	int total = 0;
	
#define SKIP_CLASS\
	while(isspace(*p)) ++p;\
	while(!isspace(*p)) ++p;

#define SKIP_ELEMENT\
	while(*p!=':') ++p;\
	++p;\
	while(isspace(*p)) ++p;\
	while(*p && !isspace(*p)) ++p;

	while(fgets(line,sizeof(line),input)!=NULL)
	{
		int i = 0;
		double label,v;
		const char *p = line;

		if(sscanf(p,"%lf",&label)!=1) break;

		SKIP_CLASS

		while(sscanf(p,"%d:%lf",&x[i].index,&x[i].value)==2)
		{
			SKIP_ELEMENT;
			++i;
		}

		x[i].index = -1;
		if(svm_classify(model,x,label,&v))
			++correct;
		++total;

		fprintf(output,"%g\n",v);
	}

	printf("correct/total = %d/%d (%g%%)\n",correct,total,
					       (double)correct/total*100);
}

int main(int argc, char **argv)
{
	FILE *input, *output;
	
	if(argc!=4)
	{
		fprintf(stderr,"usage: svm-classify test_file model_file output_file\n");
		exit(1);
	}

	input = fopen(argv[1],"r");
	if(input == NULL)
	{
		fprintf(stderr,"can't open input file %s\n",argv[1]);
		exit(1);
	}

	output = fopen(argv[3],"w");
	if(output == NULL)
	{
		fprintf(stderr,"can't open output file %s\n",argv[3]);
		exit(1);
	}

	if((model=svm_load_model(argv[2]))==0)
	{
		fprintf(stderr,"can't open model file %s\n",argv[2]);
		exit(1);
	}

	classify(input,output);
	svm_destroy_model(model);
	fclose(input);
	fclose(output);
	return 0;
}
